### DISCUSSION TITLE
SSO (OAuth2) for Clever.com leads to Invalid Credentials Page on Directus

### DISCUSSION THREAD

**Discussion by 10dimensions** [directus/directus#16160] [Archive]
Hello All,

I was trying to integrate a third party SSO provider namely Clever 
- https://www.clever.com
- https://dev.clever.com/docs/classroom-with-oauth

Have added my custom .env file below, which typically follows standard OAuth flow, namely Authorize Url, Access Url, and Profile Url to get the relevant data.
But the final redirect always leads to the following endpoint on directus. 
`admin/login?reason=INVALID_CREDENTIALS`

Is there anything to be defined on the user policy?
Any help would be greatly appreciated.

```
AUTH_PROVIDERS="clever"

AUTH_CLEVER_DRIVER="oauth2"
AUTH_CLEVER_CLIENT_ID="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
AUTH_CLEVER_CLIENT_SECRET="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
AUTH_CLEVER_AUTHORIZE_URL="https://clever.com/oauth/authorize"
AUTH_CLEVER_ACCESS_URL="https://clever.com/oauth/tokens"
AUTH_CLEVER_PROFILE_URL="https://api.clever.com/me"
AUTH_CLEVER_LABEL="Login with Clever"
AUTH_CLEVER_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_CLEVER_IDENTIFIER_KEY="email"
AUTH_CLEVER_ICON="scroll"
AUTH_CLEVER_DEFAULT_ROLE_ID="3af2c341-fe53-4dc1-9550-8f6b7d58bef5"
```

**Comment by aidenfoxx** [comment]
I would advise trying to use the OpenID driver with the config:

```
AUTH_PROVIDERS="clever"

AUTH_CLEVER_DRIVER="openid"
AUTH_CLEVER_CLIENT_ID="YOUR_ID"
AUTH_CLEVER_CLIENT_SECRET="YOUR_SECRET"
AUTH_CLEVER_ISSUER_URL="https://clever.com/.well-known/openid-configuration"
AUTH_CLEVER_IDENTIFIER_KEY="email"
AUTH_CLEVER_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_CLEVER_DEFAULT_ROLE_ID="YOUR_ROLE_ID"
```

**Reply by 10dimensions** [reply]
Thanks...Let me get back on this.
OpenID Connect needs specific permissions from Clever itself.

**Comment by 10dimensions** [comment]
Just tested our applications with OpenID configured.
This time, there's a different prompt thrown on Directus Admin CMS.

Redirecting to the following URL, with the prompt being `Unexpected Error`

`https://xxx-yyy-zzz.com/admin/login?reason=UNKNOWN_EXCEPTION`

Can you please guide as to how to debug the same? And if I'm missing something.

Thanks

**Reply by 10dimensions** [reply]
@aidenfoxx Apologies for re-notifying.
But any help would be great regarding the openID issue that I'm facing.

**Reply by aidenfoxx** [reply]
An unknown error could be anything. If you can check your logs for more details and post them here I might be able to help. You can also enable trace logging to potentially get more info.

**Reply by 10dimensions** [reply]
```
21:44:34 ✨ request completed GET 302 /auth/login/clever?redirect=https://<URL>/admin/login?continue 12ms
21:44:36 ⚠️  id_token not present in TokenSet
21:44:36 ⚠️  [OpenID] Unexpected error during OpenID login

21:44:36 ✨ request completed GET 302 /auth/login/clever/callback?code=b98e896f29828bed6534&state=HCwDULXzjMssOgbISGyLcJx7R-eGP8n-F2D7XNNyWnU 511ms
21:44:36 ✨ request completed GET 304 /admin/login?reason=UNKNOWN_EXCEPTION
```
Managed to find live logs. 
I tried to added SCOPE="openid email profile" as well to the .env

```
AUTH_CLEVER_IDENTIFIER_KEY="email"
AUTH_CLEVER_SCOPE="openid profile email"
```

But still the issue persists.
Thanks in advance @aidenfoxx .

**Reply by biapar** [reply]
Maybe the Clever reply is not in correct form.

**Reply by aidenfoxx** [reply]
As @biapar said, the Clever response does not include the `id_token` which is a requirement of the OpenID flow we use. You can go back and try and get the OAuth configuration working, or contact Clever to understand what's going wrong with their OpenID configuration.

**Reply by aidenfoxx** [reply]
Is it also possible that if you've previously authenticated your app without an OpenID scope, that your permissions are misconfigured on the Clever side. Try revoking access in Clever, and re-authenticating via Directus.
